R/timeseries.R

```diff
@@ -1,3 +1,568 @@
+                    .fns = reduction),
+      .groups = "drop"
+    )
+  )
+
+}
+
+
+#' bootstrap a `tseries` object containing original data
+#'
+#' @param .tseries `tseries` object with `tseries_orig` mixin
+#' @param boot.R Intenger, number of bootstrap samples to use.
+#' @param boot.l Integer, block length to use in bootstrap procedure.
+#' @param seed Integer, random number generator seed to be used in bootstrap.
+#' @param sim String, `sim` argument passed to \link[boot]{tsboot}.
+#' @param endcorr Boolean, `endcorr` argument passed to \link[boot]{tsboot}.
+#' @param serial Boolean, whether to disable the use of \link[parallel]{mclapply}.
+#'
+#' @return `tseries` object with the additional members
+#' \itemize{
+#' }
+#'
+#' @examples
+#' xyz_md <- expand.grid(x = 0:9, y = 0:9, z = 0:9, md_idx = 1:100)
+#' new_tseries <- tseries_orig(data = cbind(xyz_md, val = rnorm(nrow(xyz_md))),
+#'                             explanatory_vars = c("x", "y", "z") )
+#'
+#' bs_tseries <- bootstrap.tseries(new_tseries,
+#'                                 boot.R = 500, boot.l = 5, seed = 12345,
+#'                                 sim = 'geom', endcorr = TRUE, serial = FALSE)
+#'
+#' @export
+bootstrap.tseries <- function(.tseries, 
+                              boot.R, boot.l, seed, sim, endcorr,
+                              serial = FALSE) {
+
+  dplyr_version_required <- "1.0.0"
+  dplyr_avail <- requireNamespace("dplyr", versionCheck = list(op = ">=", version = dplyr_version_required))
+  if( !dplyr_avail ){
+    stop(sprintf("The 'dplyr' package in version >= %s is required to use this function!\n",
+                 dplyr_version_required))
+  }
+
+  parallel_avail <- requireNamespace("parallel")
+  if( !parallel_avail & serial == FALSE ){
+    stop(sprintf("The 'parallel' package in version is required to use this function with `serial = FALSE`!\n"))
+  }
+
+  if( serial ){
+    my_lapply <- lapply
+  } else {
+    my_lapply <- parallel::mclapply
+  }
+  
+  stopifnot(inherits(.tseries, 'tseries_orig'))
+
+  boot.l <- ceiling(boot.l)
+  boot.R <- floor(boot.R)
+
+  stopifnot(boot.l >= 1)
+  stopifnot(boot.l <= length(unique(.tseries$data$md_idx)))
+  stopifnot(boot.R >= 1)
+
+  old_seed <- swap_seed(seed)
+  # unlike in other uses of tsboot in hadron, here we bootstrap the
+  # `md_idx`
+  sample_idcs <- boot::tsboot(unique(.tseries$data$md_idx),
+                              statistic = function(x){ x },
+                              R = boot.R,
+                              l = boot.l,
+                              sim = sim,
+                              endcorr = endcorr)$t
+  restore_seed(old_seed)
+
+  ts_samples <- my_lapply(X = 1:nrow(sample_idcs),
+    FUN = function(row_idx)
+      {
+        idcs <- sample_idcs[row_idx,,drop=TRUE]
+        # compute the row indices corresponding to the bootstrap sample indices
+        df_sample_idcs <- unlist(lapply(idcs, function(x){ which(x == .tseries$data$md_idx) }))
```

Comment:
I'm afraid this approach is doomed to failure because this statement (repeated `boot.R` times) is too expensive when `.tseries$data$md_idx` is of length O(100k), which is the case for the situations of interest... Basically I think I've found the reason why high-dimensional data should not be stored in long format if it is to be bootstrapped... back to the drawing board!

Reply:
I have a workaround using the `grr` package which at least makes it bearable...